pkg-manager/core/src/install/getPreferredVersions.ts

```diff
@@ -21,15 +22,32 @@ export function getAllUniqueSpecs (manifests: DependencyManifest[]) {
   return allSpecs
 }
 
-export function getPreferredVersionsFromLockfile (snapshots: PackageSnapshots): PreferredVersions {
-  return Object.entries(snapshots)
-    .map(([depPath, snapshot]) => nameVerFromPkgSnapshot(depPath, snapshot))
-    .reduce((preferredVersions, { name, version }) => {
-      if (!preferredVersions[name]) {
-        preferredVersions[name] = { [version]: 'version' }
-      } else {
-        preferredVersions[name][version] = 'version'
-      }
-      return preferredVersions
-    }, {} as PreferredVersions)
+export function getPreferredVersionsFromLockfileAndManifests (
+  snapshots: PackageSnapshots | undefined,
+  manifests: Array<DependencyManifest | ProjectManifest>
+): PreferredVersions {
+  const preferredVersions: PreferredVersions = {}
+  for (const manifest of manifests) {
+    const specs = getAllDependenciesFromManifest(manifest)
+    for (const [name, spec] of Object.entries(specs)) {
+      const selector = getVersionSelectorType(spec)
+      if (!selector) continue
+      preferredVersions[name] = preferredVersions[name] ?? {}
+      preferredVersions[name][spec] = selector.type
+    }
+  }
+  if (!snapshots) return preferredVersions
+  getPreferredVersionsFromLockfile(snapshots, preferredVersions)
+  return preferredVersions
+}
+
+function getPreferredVersionsFromLockfile (snapshots: PackageSnapshots, preferredVersions: PreferredVersions) {
```

Comment:
nit: May make sense to call this `addPreferredVersionsFromLockfile` since it's now mutating.